### PR TITLE
Add question_type field to exported JSON

### LINKED ISSUE #5437: Add question type to JSON response
AR would like to add the question type to the JSON response, so that it looks like

```
"applicant_favorite_season": {
  "question_type": "TEXT",
  "text": "autumn"
}
```

This will require changes to the enumerator question structure as well. Currently, it looks like
```
"household_members": [
  {...},
  {...},
  {...}
]
```
but that doesn't give us anywhere to add the question type or additional metadata, so we'll restructure it to
```
"household_members": {
  "question_type": "ENUMERATOR",
  "entities": [
    {...},
    {...},
    {...}
  ]
}
```

We will use the QuestionType enum as the exported question type. This means that even though both Dropdown and Radio Button questions use the same `selection` scalar, they'll have different question types.

### PR DESCRIPTION
**Pull request by avaleske**
### Description

Adds a `question_type` field to answers in exported JSON.

Answers now include a `question_type` field, like:
```js
"applicant_favorite_season": {
  "question_type": "TEXT",
  "text": "autumn"
}
```

and answers to enumerator questions look like:
```js
"household_members": {
  "question_type": "ENUMERATOR",
  "entities": [
    {...},
    {...},
    {...}
  ]
}
```
where previously, answers to enumerator questions were not nested:
```js
"household_members": [
  {...},
  {...},
  {...}
]
```

To support this, this PR also:
- Adds a new `QuestionTypeExternal` enum with the external name of the question types. This may be distinct from the internal name, for example, both Dropdown and Radio Button questions are `SINGLE_SELECT`.
- Adds additional methods to `Path` to support building the nested path.
- Moves API specific path segments to `ApiPathSegments`.

## Release notes

Changes the format of enumerator questions in the JSON export, and adds `question_type` to all questions.

#### General

Read the full guidelines for PRs [here](https://docs.civiform.us/contributor-guide/developer-guide/technical-contribution-guide#guidelines-for-pull-requests)

- [x] Added the correct label: < feature | bug | dependencies | infrastructure | ignore-for-release | database >
- [x] Created unit and/or browser tests which fail without the change (if possible)
- [x] Performed manual testing (Chrome and Firefox if it includes front-end changes)
- [ ] Extended the README / documentation, if necessary (See #5326)

#### User visible changes

n/a

#### New Features

n/a

### Instructions for manual testing

If desired:
- Export an application as JSON, and inspect the response for the question_type field.
- In particular, check nested repeated entities.

### Issue(s) this completes

Fixes #5437

### CODE REVIEW

**Comment by avaleske** [comment]
Current dependencies on/for this PR:
* main
  * **PR #5485** <a href="https://app.graphite.dev/github/pr/civiform/civiform/5485" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/civiform/civiform/5485?utm_source=stack-comment).

File: server/app/services/export/JsonExporter.java [hunk 1]
```diff
@@ -234,6 +228,8 @@ private static void exportApplicationEntriesToJsonApplication(
         @SuppressWarnings("unchecked")
         ImmutableList<String> list = (ImmutableList<String>) maybeJsonValue.get();
         jsonApplication.putArray(path, list);
+      } else if (instanceOfEmptyImmutableList(maybeJsonValue.get())) {
+        jsonApplication.putArray(path, ImmutableList.of());
```
Comment (avaleske) [inline server/app/services/export/JsonExporter.java:232]: Restores the logic that was removed in #5306, which was the better way to do this.

File: server/app/services/Path.java [hunk 2]
```diff
@@ -170,12 +220,22 @@ public Path asArrayElement() {
...
     return parentPath().join(keyNameWithoutArrayIndex());
   }
 
+  /**
+   * A version of {@link #withoutArrayReference()} that doesn't throw an exception if there is not
+   * an array reference.
+   *
+   * @return a path with the trailing array element reference stripped away.
+   */
+  public Path safeWithoutArrayReference() {
```
Comment (avaleske) [inline server/app/services/Path.java:235]: I'm open to suggestions for names for this

File: browser-test/image_snapshots/api_docs_test/comprehensive-program-active-version.png
Comment (dkatzz) [inline browser-test/image_snapshots/api_docs_test/comprehensive-program-active-version.png:1]: **question:** Unrelated to this PR specifically, but why don't the enumerator questions show up in the preview?

File: server/app/services/Path.java [hunk 3]
```diff
@@ -118,6 +119,27 @@ public Path join(String path) {
    */
   public Path join(Scalar scalar) {
     Path other = Path.create(scalar.name());
+    return join(other);
+  }
+
+  /**
+   * Append the provided {@link ApiPathSegment} to this {@link Path}.
```
Comment (dkatzz) [inline server/app/services/Path.java:126]: **thought:** Would be nice to make the other join functions javadocs consistent (can be here or in a separate PR).

File: server/test/services/export/JsonExporterTest.java [hunk 4]
```diff
@@ -304,7 +310,12 @@ public void export_whenCheckboxQuestionIsNotAnswered_valueInResponseIsEmptyArray
             SubmittedApplicationFilter.EMPTY);
     ResultAsserter resultAsserter = new ResultAsserter(resultJsonString);
 
-    resultAsserter.assertValueAtApplicationPath(".kitchen_tools.selections", ImmutableList.of());
+    resultAsserter.assertJsonAtApplicationPath(
```
Comment (dkatzz) [inline server/test/services/export/JsonExporterTest.java:313]: **suggestion:** Another option so you don't have to worry as much about formatting would be to have `resultAsserter.assertJsonAtApplicationPath(".kitchen_tools.selections"..`
and
`resultAsserter.assertJsonAtApplicationPath(".kitchen_tools.question_type"..`

Here and in other parts of the file. Fine with either though.

**Review by dkatzz — APPROVED** [review]

File: browser-test/image_snapshots/api_docs_test/comprehensive-program-active-version.png
Comment (avaleske) [reply browser-test/image_snapshots/api_docs_test/comprehensive-program-active-version.png:1]: They're not implemented yet. That's #5238.

File: server/test/services/export/JsonExporterTest.java (hunk 4 above)
Comment (avaleske) [reply server/test/services/export/JsonExporterTest.java:313]: yeah... I wanted to start enforcing that other stuff we're not expecting doesn't show up in the API response, not just that the things we're expecting are there. But I don't like that the order matters in these asserts, and that the test failure .diffs are really hard to read

I'm planning to look into jayway.JsonAssert, which I think could be used here to make these more readable while also checking that we're not including anything we don't want to.

File: server/app/services/Path.java (hunk 3 above)
Comment (avaleske) [reply server/app/services/Path.java]: Done!

File: server/app/services/Path.java [hunk 5]
```diff
@@ -100,24 +102,47 @@ public Path replacingLastSegment(String finalSegment) {
   }
 
   /**
-   * Append a path to the path.
+   * Append a String path to the path.
```
Comment (dkatzz) [inline server/app/services/Path.java:105]: **suggestion:** Add {@link Path} for consistency here as well?

File: server/app/services/Path.java [hunk 6]
```diff
@@ -100,24 +102,47 @@ public Path replacingLastSegment(String finalSegment) {
...
+   * @param apiPathSegment the ApiPathSegment to join with this path
+   * @return this path joined with the provided ApiPathSegment
+   */
+  public Path join(ApiPathSegment apiPathSegment) {
+    Path other = Path.create(apiPathSegment.name());
+    return join(other);
+  }
+
+  /**
+   * Append a {@link Path} to this {@link Path}.
```
Comment (dkatzz) [inline server/app/services/Path.java:140]: **nitpick:** this -> the for consistency

File: browser-test/image_snapshots/api_docs_test/comprehensive-program-active-version.png
Comment (dkatzz) [reply browser-test/image_snapshots/api_docs_test/comprehensive-program-active-version.png:1]: Ah makes sense thanks!

File: server/test/services/export/JsonExporterTest.java (hunk 4 above)
Comment (dkatzz) [reply server/test/services/export/JsonExporterTest.java:313]: Sounds good, makes sense!

**Review by dkatzz — APPROVED** [review]

File: server/app/services/Path.java (hunk 5 above)
Comment (avaleske) [reply server/app/services/Path.java]: Done

File: server/app/services/Path.java (hunk 6 above)
Comment (avaleske) [reply server/app/services/Path.java]: Done.

**Comment by avaleske** [comment]
FYI @gwendolyngoetz
